File: 主成分分析/主成分分析(PCA).py
#主成分分析(PCA)
#無監督(不需要用到應變數)
#套件
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from matplotlib.colors import ListedColormap

#read
data = pd.read_csv('Wine.csv')
x = data.iloc[:, :13].values
y = data.iloc[:, 13].values

#train、test
x_train, x_test, y_train, y_test = train_test_split(
        x, y, test_size = 0.2)

#feature scaling
sc_x = StandardScaler()
x_train = sc_x.fit_transform(x_train)
x_test = sc_x.transform(x_test) #前面fit過了

#apply PCA
pca = PCA(n_components = 2) #最後提取兩個自變數
x_train = pca.fit_transform(x_train)
x_test = pca.transform(x_test) #前面fit過了
explained_variance = pca.explained_variance_ratio_

#邏輯迴歸
model = LogisticRegression()
model.fit(x_train, y_train)

#predict
y_pred = model.predict(x_test)

#confusion_matrix
cm = confusion_matrix(y_test, y_pred)

#visual(train_set)
x_set, y_set = x_train, y_train
#上下左右都留白一格單位
#網格
x1, x2 = np.meshgrid(np.arange(x_set[:, 0].min() - 1, x_set[:, 1].max() + 1,
                               0.01),
                     np.arange(x_set[:, 1].min() - 1, x_set[:, 1].max() + 1,
                               0.01))
#分界線，上色
plt.contourf(x1, x2, model.predict(
        np.array([x1.ravel(), x2.ravel()]).T).reshape(x1.shape),
        alpha = 0.75, cmap = ListedColormap(('red', 'green', 'black')))
plt.xlim(x1.min(), x1.max())
plt.ylim(x2.min(), x2.max())

#將點上色
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(x_set[y_set == j, 0], x_set[y_set == j, 1],
                c = ListedColormap(('orange', 'blue', 'grey'))(i),
                label = j)

plt.title('Logistic Regression(train_set)')
plt.xlabel('PC1')
plt.ylabel('PC2')
plt.legend()
plt.show()

#test_set
x_set, y_set = x_test, y_test
#上下左右都留白一格單位
#網格
x1, x2 = np.meshgrid(np.arange(x_set[:, 0].min() - 1, x_set[:, 1].max() + 1,
                               0.01),
                     np.arange(x_set[:, 1].min() - 1, x_set[:, 1].max() + 1,
                               0.01))
#分界線，上色
plt.contourf(x1, x2, model.predict(
        np.array([x1.ravel(), x2.ravel()]).T).reshape(x1.shape),
        alpha = 0.75, cmap = ListedColormap(('red', 'green', 'black')))
plt.xlim(x1.min(), x1.max())
plt.ylim(x2.min(), x2.max())

#將點上色
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(x_set[y_set == j, 0], x_set[y_set == j, 1],
                c = ListedColormap(('orange', 'blue', 'grey'))(i),
                label = j)

plt.title('Logistic Regression(test_set)')
plt.xlabel('PC1')
plt.ylabel('PC2')
plt.legend()
plt.show()

File: 主成分分析/主成分分析(PCA).r
#主成分分析(PCA)
#不須使用應變數(無監督)
#read
data = read.csv('Wine.csv')

#train、test
library(caTools)
split = sample.split(data$Customer_Segment, SplitRatio = 0.8)
train_set = subset(data, split == TRUE)
test_set = subset(data, split == FALSE)

#feature scaling
train_set[,-14] = scale(train_set[, -14])
test_set[, -14] = scale(test_set[, -14])

#apply PCA
library(caret)
library(e1071)
#pcaComp表示最後提取出的自變數數量
pca = preProcess(x = train_set[, -14], method = 'pca', pcaComp = 2)
train_set = predict(pca, train_set)
train_set = train_set[c(2, 3, 1)] #欄位更換
test_set = predict(pca, test_set)
test_set = test_set[c(2, 3, 1)] #欄位更換

#SVM model
model = svm(formula = Customer_Segment ~ .,
            data = train_set,
            type = 'C-classification',
            kernel = 'linear')

#predict
y_pred = predict(model, test_set[, -3])

#confusion matrix
cm = table(test_set[, 3], y_pred)

#visual(train_set)
library(ElemStatLearn)
set = train_set
#上下左右留白一個單位
x1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.0075)
x2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.0075)
#網格
grid_set = expand.grid(x1, x2)
colnames(grid_set) = c('PC1', 'PC2')
#預測網格
y_grid = predict(model, grid_set)
#畫點
plot(set[, -3],
     main = 'model', xlab('PC1'), ylab('PC2'),
     xlim = range(x1), ylim = range(x2))
#分界線
contour(x1, x2, matrix(as.numeric(y_grid), length(x1), length(x2)), add = TRUE)
#將區域上色
points(grid_set, pch = '.', col = ifelse(y_grid == 2, 'Deepskyblue',
                          ifelse(y_grid == 1, 'springgreen3', 'tomato')))
#將點上色
points(set, pch = 21, bg = ifelse(set[, 3] == 2, 'blue3', 
                          ifelse(set[, 3] == 1, 'green4', 'red3')))

#test_set
library(ElemStatLearn)
set = test_set
#上下左右留白一個單位
x1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.0075)
x2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.0075)
#網格
grid_set = expand.grid(x1, x2)
colnames(grid_set) = c('PC1', 'PC2')
#預測網格
y_grid = predict(model, grid_set)
#畫點
plot(set[, -3],
     main = 'model', xlab('PC1'), ylab('PC2'),
     xlim = range(x1), ylim = range(x2))
#分界線
contour(x1, x2, matrix(as.numeric(y_grid), length(x1), length(x2)), add = TRUE)
#將區域上色
points(grid_set, pch = '.', col = ifelse(y_grid == 2, 'Deepskyblue',
                                         ifelse(y_grid == 1, 'springgreen3', 'tomato')))
#將點上色
points(set, pch = 21, bg = ifelse(set[, 3] == 2, 'blue3', 
                                  ifelse(set[, 3] == 1, 'green4', 'red3')))
